### DISCUSSION TITLE
Issues using node playground with custom message types

### DISCUSSION THREAD

**Discussion by jackiescanlon** [foxglove/community#56] [Q&A] [answered]
I've been using a few Node Playground panels to convert units. But, I'm having an issue using custom message types.

Here's a simplified example:

## Setup
My custom message type is `merlin_common/Float32Stamped`, which looks like this:
```
Header header
float32 value
```
I'm running a simple node that subscribes to a `Float32Stamped` message in meters, converts it to feet, and publishes another `Float32Stamped` via ROS 1 live connection.

Here's my studio node code:

```ts
import { Input, Messages } from "ros";

export const M_TO_FT = 3.28084;

type Output = Messages.merlin_common__Float32Stamped;

export const inputs = ["/sine_wave_stamped"];
export const output = "/studio_node/sine_wave_stamped_ft";

export default function node(event: Input<"/sine_wave_stamped">): Output {
  return {
    header: event.message.header,
    value: event.message.value * M_TO_FT,
  };
}
```

My layout is `sine_wave.json`. The first tab is a plot panel, plotting `/studio_node/sine_wave_stamped_ft.value`. The second tab includes the studio node.

`sine_wave.json` as well as a test bag file with recorded data, `sine_wave.bag`, are included in 
[sine_wave.zip](https://github.com/foxglove/studio/files/8909478/sine_wave.zip).

## Testing
Loading it with the live connection, the legend text is red, and nothing is plotting:

![image](https://user-images.githubusercontent.com/43035541/173829707-53250b87-dbb2-424d-a358-ca7b09649bfe.png)

Clicking over to Tab 2, I can see these "Problems":

![image](https://user-images.githubusercontent.com/43035541/173829754-edbf97fa-7ceb-49f7-9cd9-56db087d3db4.png)

That doesn't make a lot of sense, I know that `Float32Stamped` includes both `header` and `value` properties.

## Workaround
Then, I figured out something weird: If I add another plot that plots the original topic `/sine_wave_stamped`, suddenly Foxglove studio clears the "Problems" and is able to plot the new topic just fine:

![image](https://user-images.githubusercontent.com/43035541/173830019-29694c51-2594-4351-abfc-9f9df2bb3ac3.png)

![image](https://user-images.githubusercontent.com/43035541/173830143-f21f0780-6655-4af8-bc02-8c558318274a.png)

It's almost as if Studio can't figure out what a `Float32Stamped` is, but once it receives a message of that type, it figures it out.

Also, if I save the layout on Tab 1 and re-open it later, it won't work again. I have to click over to Tab 2, where the bonus plot is, and let it plot a point before it will resolve the issue again.

## Help

Is there something I can change about my node code or my Foxglove Studio configuration to make it so that I don't need a plot to help Foxglove find the message types?

**Comment by defunctzombie** [comment]
Few things I'm noticing:

- Your code snipped doesn't match what is in your screenshot code snippets. The screenshots have an output type that extends another type.
- The "ros" import is considered deprecated for node playground code - and a more robust `Message` and `Input` generic type are available from `./types`

Give the following code a try and see if that works or post any errors you see when trying to use this code.

```typescript
import { Input, Message } from "./types";

export const M_TO_FT = 3.28084;

type Output = Message<"merlin_common/Float32Stamped">;

export const inputs = ["/sine_wave_stamped"];
export const output = "/studio_node/sine_wave_stamped_ft";

export default function node(event: Input<"/sine_wave_stamped">): Output {
  return {
    header: event.message.header,
    value: event.message.value * M_TO_FT,
  };
}
```

**Reply by jackiescanlon** [reply]
Oops, sorry! I  was trying something else (that didn't work) and never reverted my code for the screenshot.

Here's the output for the code snippet you sent. Still getting 2 errors, albeit different ones.

![image](https://user-images.githubusercontent.com/43035541/173849510-c91f8eb9-86ce-4754-983f-50e4c86367b7.png)

Beyond that error though, I've had numerous issues using `./types` before:

* When I run my full system (with many topics and many custom messages) and not just the example code here, these are the errors I get using `ros`. I only get the 2 errors I normally do, and am able to workaround with the extra plot.

![image](https://user-images.githubusercontent.com/43035541/173852392-22fc181c-5ebb-45db-bdab-2780918818d0.png)

* When I run my full system and use `./types`, the node throws many many errors besides these two. My workaround of plotting a message of the correct type doesn't work.

![image](https://user-images.githubusercontent.com/43035541/173851515-f2e3c3d8-ef47-44d4-a6c2-469798cf0900.png)

**Reply by defunctzombie** [reply]
This last screenshot looks like you are running into a bug of some sort that prevents your custom datatype types from being available in node playground. I've seen similar before so would like to see if I can reproduce this and find a fix.

Above you mentioned this issue happening when using the ROS1 native connection. Does it also happen if you load a bag file with your data?

**Reply by jackiescanlon** [reply]
Ah yeah that sounds like it.

Re: bag file:

For my little example, I recorded `sine_wave_stamped.bag` attached [here](https://github.com/foxglove/studio/files/8948709/sine_wave_stamped.zip). When I load that into Foxglove and use your code snippet (with `./types`) there's no issue, it plots the data right away without needing any workarounds. Also works just fine when I record data from my full system as well.

**Comment by defunctzombie** [comment, marked as answer]
@jackiescanlon Thank you for the issue report. I've reproduced the issue and created a fix here: https://github.com/foxglove/studio/pull/3700

Once this lands in a release (early next week), you'll be able to use your custom datatypes in node playground scripts.

**Reply by jackiescanlon** [reply]
Woohoo!! Thanks you so much!

**Reply by jackiescanlon** [reply]
I tested this in 1.16.0 but still getting the same errors as before with `./types`. It looks like 1.16.0 was created after your PR #3700 was merged, but I don't see it in the release notes - so maybe its going to be in a future release?

**Reply by jtbandes** [reply]
That PR was included in the 1.16.0 release – the surefire way to know is to look here on https://github.com/foxglove/studio/commit/be955870bad4bff07fc3d72ef246a88296cdc874: 
<img width="141" alt="image" src="https://user-images.githubusercontent.com/14237/177622601-f55c5c59-fd3c-4ac6-887f-a0935108173d.png">

**Reply by jackiescanlon** [reply]
Ah thanks!

Unfortunately that means that my issue is still happening :(

**Reply by defunctzombie** [reply]
@jackiescanlon To confirm - you've opened your file and the error remains? If you are still seeing this could you produce a sample file and attach it to a new github bug report. I was able to reproduce and debug this in the linked PR but maybe there's something different about the file you are loading. Please tag me on the issue.

**Reply by jackiescanlon** [reply]
Sounds good, thanks! To clarify, the simple `sine_wave` example that I used seems to work fine. But I'm still running into errors when I run my full system, with the node playground unable to resolve other custom message types that are irrelevant to the node playground code.

I'll come up with another example that uses multiple custom message types so that you can try to replicate it.

**Reply by jackiescanlon** [reply]
https://github.com/foxglove/studio/issues/3784
